docs/architecture/adr-013-metrics.md

```diff
@@ -0,0 +1,90 @@
+# ADR 013: Observability
+
+## Changelog
+
+- 20-01-2020: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+There has been discussion around exposing more metrics to users and node operators about the application. Currently there is only a way to expose metrics from Tendermint and not the application itself. To bring more visibility into applications, I would like to propose reporting of metrics through [Prometheus](https://prometheus.io/).
+
+Extending `AppModuleBasic` to support registering of metrics would enable developers to see more information about individual modules.
+
+```go
+type AppModuleBasic interface {
+  Name() string
+  RegisterCodec(*codec.Codec)
+  RegisterMetrics(namespace string, labelsAndValues ...string) *Metrics
+
+  // genesis
+  DefaultGenesis() json.RawMessage
+  ValidateGenesis(json.RawMessage) error
+
+  // client functionality
+  RegisterRESTRoutes(context.CLIContext, *mux.Router)
+  GetTxCmd(*codec.Codec) *cobra.Command
+  GetQueryCmd(*codec.Codec) *cobra.Command
+}
+// .....
+
+func (bm BasicManager) RegisterMetrics(appName) MetricsProvider {
+	for _, b := range bm {
+		b.RegisterMetrics(appName, labelsAndValues)
+	}
+}
+```
+
+Each module can define its own `CreateMetrics` function:
+
+```go
+type Metrics struct {
+  Size metrics.Guage
+
+  Transactions metrics.Counter
+}
+
+func CreateMetrics(namespace string, labelsAndValues ...string) *Metrics {
```

Comment:
the  `CreateMetrics` should be generic so that it can be called by the module manager. In this case, you are creating a function only for the specific case of the `Size` field. Say you want to initialize also the `Counter` for `Transactions`, with the following function:
```go
NewCounterFrom(opts prometheus.CounterOpts, labelNames []string) *Counter
```

How do you know which`labelsAndValues` correspond to which metric?  This also applies for `Histogram` and `Summary`

Reply:
I believe this comes from what Tendermint did and we could change this for sure, but in TM the same labels are being added to every metric that the module wants to expose.

So the workflow is that module owner will just add Metrics they want to track in the metric struct, and to add more metrics you simple add more fields in that struct, and not only tied to one metric per say.

So if we want the flexibility that every metric can have some custom labels that gets added when pulled from Prometheus then we need to perhaps create a different API for module to define these metrics, since you need the metric type, name and the labels associated with it, potentially even global labels that gets applied to every metric like what we have here.

@fedekunze do you know if we need to have this flexibility to definte custom labels on each metrics in the same module? Or having it apply to every metric in that module is good enough?